prelude/toolchains/conan/defs.bzl

```diff
@@ -0,0 +1,791 @@
+        # cannot both be passed to `prebuilt_cxx_library`'s `include_dirs`.
+        # This adds a counter prefix to avoid the overlap.
+        prefix = str(i) + "/"
+        i += 1
+        output = ctx.actions.declare_output(prefix + dirname)
+        cmd.add(["--directory-from", dirname, "--directory-to", output.as_output()])
+        if dirname in sub_targets:
+            fail("Directory-name collision: " + dirname)
+        sub_targets[dirname] = [DefaultInfo(default_outputs = [output])]
+
+    cmd.add(["--package", ctx.attrs.package[ConanPackageInfo].package_out])
+    ctx.actions.run(cmd, category = "conan_extract")
+
+    return [DefaultInfo(default_outputs = [], sub_targets = sub_targets)]
+
+_conan_package_extract = rule(
+    impl = _conan_package_extract_impl,
+    attrs = {
+        "package": attrs.dep(providers = [ConanPackageInfo], doc = "The Conan package directory to extract files from."),
+        "files": attrs.list(attrs.string(), doc = "Files to extract from the package."),
+        "directories": attrs.list(attrs.string(), doc = "Directories to extract from the package."),
+        "_conan_package_extract": attrs.dep(providers = [RunInfo], default = "prelude//toolchains/conan:conan_package_extract"),
+    },
+    doc = "Extract files and directories from Conan package directory.",
+)
+
+def conan_component(
+        name: "string",
+        defines: ["string"],
+        cflags: ["string"],
+        cppflags: ["string"],
+        include_paths: ["string"],
+        libs: ["string"],
+        static_libs: {"string": ["string"]},
+        shared_libs: {"string": ["string"]},
+        system_libs: ["string"],
+        deps: ["string"],
+        package: "string"):
+    """Import a Conan package component.
+
+    Extracts the relevant files from the Conan package directory and exposes
+    them as a target that can be depended on by native Buck2 C/C++ targets such
+    as `cxx_library`.
+    """
+
+    extract_name = name + "_extract"
+    extract_tpl = ":" + extract_name + "[{}]"
+    extract_include_paths = [extract_tpl.format(p) for p in include_paths]
+    extract_shared_libs = { name: [extract_tpl.format(lib) for lib in libs] for name, libs in shared_libs.items() }
+    extract_static_libs = { name: [extract_tpl.format(lib) for lib in libs] for name, libs in static_libs.items() }
+
+    _conan_package_extract(
+        name = extract_name,
+        package = package,
+        files = flatten(static_libs.values() + shared_libs.values()),
+        directories = include_paths,
+    )
+
+    if len(libs) == 1:
+        lib = libs[0]
+        if lib in shared_libs:
+            shared_lib = extract_shared_libs[lib][0]
+        else:
+            shared_lib = None
+        if lib in static_libs:
+            static_lib = extract_static_libs[lib][0]
+        else:
+            static_lib = None
+        native.prebuilt_cxx_library(
+            name = name,
+            deps = deps,  # TODO[AH] Do we need exported_deps?
```

Comment:
is exported_deps not needed for transitive deps to work correctly? @ndmitchell please confirm

Reply:
For c++, exported_deps is required when your dependents need to include your dependency's headers. This could be because your headers that your dependencies include then further directly include ones from your dependency's or if you just otherwise expect your dependents to do so (like if your interfaces are going to require things from your dependency's). Often this isn't required.

Now, when looking at doing something like we're doing here, we're kinda limited by only being able to understand what the underlying specification encodes (and how its other interpretations behave). If the specification doesn't distinguish between public and private dependencies, it is almost certainly treating them all as public (it could be treating them as private, but then everything we sort of need to individually specify their transitive deps and that is rather gross). In that case, I'd recommend encoding them as public deps (i.e. exported_deps) in buck2 as well.

Reply:
@cjhopman Thanks for the context! I've changed the code to set `exported_deps` instead.